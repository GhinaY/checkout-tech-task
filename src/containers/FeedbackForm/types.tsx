import { FieldError, ValidationRule } from 'react-hook-form';
import { FeedbackSubmission } from '../../types';
import { TextFieldProps } from '@mui/material/TextField';

export type ControlledTextFieldProps = TextFieldProps  & {
  name: keyof FeedbackSubmission,
  fieldError: FieldError | undefined,
  helperText?: string,
  required?: boolean,
  maxLength?: number,
  additionalRules?: {[key: string]: ValidationRule },
};
