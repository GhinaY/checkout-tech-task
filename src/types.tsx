export type FeedbackSubmission = {
  name: string,
  email: string,
  rating: number,
  comment: string,
};

export type RatingsDistribution = {
  [key: number]: number,
};
