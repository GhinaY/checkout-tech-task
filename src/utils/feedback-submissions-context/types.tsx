import { FeedbackSubmission, RatingsDistribution } from '../../types'

export type FeedbackSubmissionsContextType = {
  addNewSubmission: (newSubmission: FeedbackSubmission) => void,
  ratingsDistribution: RatingsDistribution,
  submissions: FeedbackSubmission[],
}
