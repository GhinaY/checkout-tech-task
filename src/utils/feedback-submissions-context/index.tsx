import React, { createContext, useState } from 'react';
import { FeedbackSubmission, RatingsDistribution } from '../../types';
import { FeedbackSubmissionsContextType } from './types';

export const FeedbackSubmissionsContext = createContext<FeedbackSubmissionsContextType>(null!);

const initialRatingsDistribution: RatingsDistribution = [1,2,3,4,5].reduce((obj, rating) => {
  return {
    ...obj,
    [rating]: 0,
  };
}, {});

interface Props {
  children: React.ReactNode;
}

export const FeedbackSubmissionsContextProvider: React.FC<Props> = ({ children }: Props) => {
  const [submissions, setSubmissions] = useState<FeedbackSubmission[]>([]);
  const [ratingsDistribution, setRatingsDistribution] = useState(initialRatingsDistribution);

  // adds the new submission to the list and increases the ratingsDistribution counters as needed
  const addNewSubmission = (newSubmission: FeedbackSubmission) => {
    // in a real production app, this would more realistically be an API call to save to a real database
    // for this purely-frontend app though, a context works fine
    setSubmissions(submissions.concat(newSubmission));

    // increase the counter for the rating submitted by 1
    const newRatingValue = newSubmission.rating;
    const newRatingsDistribution = {
      ...ratingsDistribution,
      [newRatingValue]: (ratingsDistribution[newRatingValue] ?? 0) + 1,
    }
    setRatingsDistribution(newRatingsDistribution)
  };

  return (
    <FeedbackSubmissionsContext.Provider value={{ addNewSubmission, ratingsDistribution, submissions }}>
      {children}
    </FeedbackSubmissionsContext.Provider>
  );
};
