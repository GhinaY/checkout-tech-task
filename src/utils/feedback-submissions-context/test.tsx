/* eslint-disable react-refresh/only-export-components */
import { useContext } from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { FeedbackSubmissionsContext, FeedbackSubmissionsContextProvider } from './index'

const mockSubmission = {
  name: 'test name',
  email: 'test@test.com',
  rating: 5,
  comment: 'test comment',
}

const TestConsumer = () => {
  const { 
    addNewSubmission, 
    ratingsDistribution,
    submissions,
  } = useContext(FeedbackSubmissionsContext);

  return (
    <>
      Number of submissions: {submissions.length}
      {submissions.map((submission, i) => (
          <span key={i}>{submission.email}: {submission.comment}</span>
      ))}
      {Object.entries(ratingsDistribution).map(
        ([rating, count]) => (
            <span key={rating}>{rating}: {count}</span>
        )
      )}
      <button onClick={() => addNewSubmission(mockSubmission)}>
          Add new submission
      </button>
    </>
  )
}
describe('FeedbackSubmissionsContextProvider', () => {
  beforeEach(() => {
    render(
      <FeedbackSubmissionsContextProvider>
        <TestConsumer />
      </FeedbackSubmissionsContextProvider>
    );
  });

  test('Provides initial ratingsDistribution correctly', () => {  
    const text = screen.getByText(
      /1: 0/i
    );
    expect(text.textContent).toBeTruthy()
  })

  test('Provides initial submissions list correctly', () => {  
    const text = screen.getByText(
      /Number of submissions: 0/i
    );
    expect(text.textContent).toBeTruthy()
  })

  describe('addNewSubmission value', () => {
    test('updates ratingsDistribution values', async () => {  
      const initialText = screen.getByText(
        /5: 0/i
      );
      expect(initialText.textContent).toBeTruthy()

      fireEvent.click(screen.getByText(/Add new submission/i));

      const updatedText = screen.getByText(
        /5: 1/i
      );
      
      expect(updatedText.textContent).toBeTruthy();
    })

    test('updates submissions values', () => {  
      const initialListLengthText = screen.getByText(
        /Number of submissions: 0/i
      );

      expect(initialListLengthText.textContent).toBeTruthy()

      fireEvent.click(screen.getByText(/Add new submission/i));

      const updatedListLengthText = screen.getByText(
        /Number of submissions: 1/i
      );
      const updatedListValueText = screen.getByText(
        /test@test.com: test comment/i
      );

      expect(updatedListLengthText.textContent).toBeTruthy()
      expect(updatedListValueText.textContent).toBeTruthy()
    })
  });
});
