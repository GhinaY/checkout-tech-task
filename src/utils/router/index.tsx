import { Navigate, createBrowserRouter } from 'react-router-dom';
import FormPage from '../../pages/FormPage';
import ResultsPage from '../../pages/ResultsPage';

export const paths = [
  {
    path: '/',
    element: <FormPage />,
  },
  {
    path: '/results',
    element: <ResultsPage />,
  },
  { 
    path: '*', 
    element: <Navigate to='/' replace /> },
]

const router = createBrowserRouter(paths);

export default router;
