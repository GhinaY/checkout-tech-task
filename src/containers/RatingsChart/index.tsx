import { useContext } from 'react';
import { Bar, BarChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Card, SxProps, Typography } from '@mui/material';
import { FeedbackSubmissionsContext } from '../../utils/feedback-submissions-context';

export const PLACEHOLDER_TEXT = 'Hmm... Looks like there are no ratings to chart';

type ContainerStylingProps = SxProps;

function RatingsChart(containerStylingProps: ContainerStylingProps) {
  const { submissions, ratingsDistribution } = useContext(FeedbackSubmissionsContext);

  const formattedRatingsDistribution = Object.entries(ratingsDistribution).map(([rating, count]) => ({
    rating: `${rating} star`, 
    count
  }));
  
  return (
    <Card variant='outlined' sx={{
      p: 1,  
      minHeight: '100px',
      display: 'flex', 
      alignItems: 'center', 
      justifyContent: 'center', 
      boxSizing: 'border-box',
      ...containerStylingProps // this allows the parent to style this container as needed
    }}>
      {/* ^ Default Card styling is just 'to taste', keeping responsiveness in mind */}

      {/* Added a placeholder message for visibility. 
      i.e. to keep the user informed of why there's no chart shown */}
      {submissions.length ? (
        <ResponsiveContainer aspect={1.5} data-testid='chart-container'>
          <BarChart data={formattedRatingsDistribution}>
            <XAxis dataKey='rating' />
            <YAxis />
            <Tooltip />
            <Bar dataKey='count' fill='#82ca9d' />
          </BarChart>
        </ResponsiveContainer>
      ) : (
        <Typography align='center'>
          {PLACEHOLDER_TEXT}
        </Typography>
      )}
    </Card>
  );
}

export default RatingsChart;
