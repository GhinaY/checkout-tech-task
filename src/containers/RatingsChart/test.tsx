/* eslint-disable react-refresh/only-export-components */
import { render, screen } from '@testing-library/react'
import { FeedbackSubmissionsContext } from '../../utils/feedback-submissions-context';
import { FeedbackSubmission, RatingsDistribution } from '../../types';
import { PLACEHOLDER_TEXT } from './index'
import RatingsChart from './index';
import { BarChart, ResponsiveContainerProps } from 'recharts';

vi.mock('recharts', async () => {
  const actual = await vi.importActual('recharts');
  return {
    ...actual!,
    ResponsiveContainer: (props: ResponsiveContainerProps) => <div>{props.children}</div>,
    BarChart: vi.fn().mockImplementation(() => <div>BarChart component</div>),
  }
})

const mockBarChart = vi.mocked(BarChart);

const mockSubmission: FeedbackSubmission = {
  name: 'test name',
  email: 'test@test.com',
  rating: 5,
  comment: 'test comment',
};

const mockRatingsDistribution: RatingsDistribution = {
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
};

describe('RatingsChart', () => {
  const mockSubmissionsList = Array(10).fill(mockSubmission);

  const renderComponent = (
    submissions: FeedbackSubmission[], 
    ratingsDistribution: RatingsDistribution
  ) => {
    return render(
      <FeedbackSubmissionsContext.Provider value={{ 
        submissions,
        ratingsDistribution,
        addNewSubmission: vi.fn() 
      }}>
        <RatingsChart />
      </FeedbackSubmissionsContext.Provider>
    );
  };

  afterEach(() =>{
    mockBarChart.mockClear()
  });

  test('Matches screenshot', () => {
    const { asFragment } = renderComponent(mockSubmissionsList, mockRatingsDistribution);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Transforms the data share correctly', () => {
    renderComponent(mockSubmissionsList, mockRatingsDistribution);

    expect(mockBarChart).toHaveBeenLastCalledWith(expect.objectContaining({ 
      data: [
        {
          rating: '1 star',
          count: 1
        },
        {
          rating: '2 star',
          count: 2
        },{
          rating: '3 star',
          count: 3
        },{
          rating: '4 star',
          count: 4
        },{
          rating: '5 star',
          count: 5
        },
      ]
    } ), expect.anything());
  });

  describe('If no submissions are available', () => {
    test('Does not render chart components', () => {
      renderComponent([], mockRatingsDistribution);
      const result = screen.queryAllByTestId('chart-container');
      expect(result.length).toBe(0)
    });

    test('Shows placeholder text', () => {
      renderComponent([], mockRatingsDistribution);
      const result = screen.getByText(PLACEHOLDER_TEXT);
      
      expect(result.textContent).toBeTruthy();
    });
  })
});
