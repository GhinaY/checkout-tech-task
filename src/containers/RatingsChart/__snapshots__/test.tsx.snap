// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RatingsChart > Matches screenshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1t6y9n7-MuiPaper-root-MuiCard-root"
  >
    <div>
      <div>
        BarChart component
      </div>
    </div>
  </div>
</DocumentFragment>
`;
